influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -0,0 +1,804 @@
+    async fn delete(&self, location: &Path) -> ObjectStoreResult<()> {
+        let result = self.inner.delete(location).await?;
+        self.delete_cache_value(location).await;
+        Ok(result)
+    }
+
+    fn delete_stream<'a>(
+        &'a self,
+        locations: BoxStream<'a, ObjectStoreResult<Path>>,
+    ) -> BoxStream<'a, ObjectStoreResult<Path>> {
+        locations
+            .and_then(|_| futures::future::err(Error::NotImplemented))
+            .boxed()
+    }
+
+    fn list(&self, prefix: Option<&Path>) -> BoxStream<'_, ObjectStoreResult<ObjectMeta>> {
+        self.inner.list(prefix)
+    }
+
+    fn list_with_offset(
+        &self,
+        prefix: Option<&Path>,
+        offset: &Path,
+    ) -> BoxStream<'_, ObjectStoreResult<ObjectMeta>> {
+        self.inner.list_with_offset(prefix, offset)
+    }
+
+    async fn list_with_delimiter(&self, prefix: Option<&Path>) -> ObjectStoreResult<ListResult> {
+        self.inner.list_with_delimiter(prefix).await
+    }
+
+    async fn copy(&self, from: &Path, to: &Path) -> ObjectStoreResult<()> {
+        self.inner.copy(from, to).await
+    }
+
+    async fn rename(&self, from: &Path, to: &Path) -> ObjectStoreResult<()> {
+        self.inner.rename(from, to).await
+    }
+
+    async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> ObjectStoreResult<()> {
+        self.inner.copy_if_not_exists(from, to).await
+    }
+
+    async fn rename_if_not_exists(&self, from: &Path, to: &Path) -> ObjectStoreResult<()> {
+        self.inner.rename_if_not_exists(from, to).await
+    }
+}
+
+/// Handle [`CacheRequest`]s in a background task
+///
+/// This waits on the given `Receiver` for new cache requests to be registered, i.e., via the oracle.
+/// If a cache request is received for an entry that has not already been fetched successfully, or
+/// one that is in the process of being fetched, then this will spin a separate background task to
+/// fetch the object from object store and update the cache. This is so that cache requests can be
+/// handled in parallel.
+fn background_cache_request_handler(
+    mem_store: Arc<MemCachedObjectStore>,
+    mut rx: Receiver<CacheRequest>,
+) -> tokio::task::JoinHandle<()> {
+    tokio::spawn(async move {
+        while let Some(CacheRequest { path, notifier }) = rx.recv().await {
+            // Check that the cache does not already contain an entry for the provide path, or that
+            // it is not already in the process of fetching the given path:
+            let mut cache_lock = mem_store.cache.lock().await;
+            if cache_lock
+                .get(&path)
+                .is_some_and(|entry| entry.is_fetching() || entry.is_success())
+            {
+                continue;
+            }
+            // Put a `Fetching` state in the entry to prevent concurrent requests to the same path:
+            let _ = cache_lock.put_with_weight(path.clone(), CacheEntry::Fetching);
```

Comment:
small thing, but I'd clone the path before you acquire the lock.

Reply:
Addressed in https://github.com/influxdata/influxdb/pull/25377/commits/ac8d7d3ba9ce25cf0dd04b4471e955e8ae4e1790